app/controllers/users_controller.rb

```diff
@@ -65,6 +65,48 @@ def new
       :form_action => %w[accounts.google.com *.facebook.com login.live.com github.com meta.wikimedia.org]
     )
 
+    if current_user
+      # The user is logged in already, so don't show them the signup
+      # page, instead send them to the home page
+      redirect_to @referer || { :controller => "site", :action => "index" }
+    else
+      check_signup_allowed
+
+      self.current_user = User.new
+    end
+  end
+
+  def create
+    self.current_user = User.new(user_params)
+
+    if check_signup_allowed(current_user.email)
+      session[:referer] = safe_referer(params[:referer]) if params[:referer]
+
+      Rails.logger.info "create: #{session[:referer]}"
+
+      if current_user.invalid?
+        # Something is wrong with a new user, so rerender the form
+        render :action => "new"
+      else
+        # Save the user record
+        session[:new_user] = current_user.attributes.slice("email", "display_name", "pass_crypt", "consider_pd")
+        save_new_user
+      end
+    end
+  end
+
+  def new_association
```

Comment:
One way to do this would be to merge those two actions back into one, but that would make merged erb template quite complex, maybe based on the presence of @auth_uid hidden field.

Reply:
Ok, so I merged back two actions into one. This results in less repetitions in controller actions and the string resources at the price of somewhat complex `users/new.html.erb`,